README
This is the README file for the doctest-tools package.

To install this package, download and unpack the source package and then:

sudo python setup.py install

If you have multiple versions of python installed, you can run install once
for each version:

sudo python2.5 setup.py install
sudo python setup.py install    # for example, if 'python' is the 2.6 version
sudo python3.1 setup.py install

These will install two programs: testdoc.py and testall.py onto your system
for each version of python.  The python version number is appended to each
name: testdoc_25.py, testdoc_26.py, testdoc_31.py and the same for testall.
In addition, the version installed with just 'python' (with no version suffix)
will install testdoc.py and testall.py without version suffixes (in addition to
copies with the version suffix matching your default 'python' program).

The testdoc.py program runs the doctests in one file.  If the file passed ends
in .py, pyc or .pyo, it runs doctest.testmod, otherwise it runs
doctest.testfile.  In ether case it determines the proper directories to
add to the "Python path" so that the tests will find the proper versions of
needed modules.  Also when testing files that are not Python modules, it does
an os.chdir to the directory containing the file.

The testall.py program recursively looks for files with .py, .tst or .txt
suffixes (or the list of suffixes you provide to it) and runs testdoc.py on
each file in a separate process (so that tests do not cross contaminate each
other).  A final summary of the number of files, tests and errors is printed
along with a list of all of the files that had errors.

These programs are covered under the "MIT License" (see the LICENSE file
included in the source directory).



RELEASE_NOTES.txt
Release 1.0a3, Mar 12, 2010
===========================

Changed how setpath works.  It now adds multiple directories to the path, and
returns a list.  Each directory going back up the path that is not a package
directory, but contains package directories is added to the path now.  If the
bottom-most directory is not a package directory, it is added to the path too.

Also:

    - Changed setpath.find_root to setpath.find_roots.
    - Applied patch from issue #2.
    - Allow blank lines in testall.config files.


Release 1.0a2, Nov 9, 2009
==========================

Fixed bug #1: Does not work on Windows

Also:

  - Added .py suffixes to testall and testdoc scripts so that Windows
    recognizes them as Python scripts.
  - Changed the version info appended to the name testall and testdoc to
    testall_26.py to make the names legal Python identifiers.
  - Added find_root function in doctest_tools.setpath to find the root
    directory without putting it on sys.path.

You should remove the testall, testdoc, testall-x.y and testdoc-x.y scripts
from your system when upgrading to this release.


doctest_tools/setpath.py
# setpath.py

import os.path
import sys
import glob

def is_package(path):
    r"""Tests whether path is a Python package directory.

    The way it determines this is by checking for an __init__.py* file
    in the directory.
    """
    for _ in glob.iglob(os.path.join(path, '__init__.py*')):
        return True
    return False

def has_package(path):
    r"""Tests whether path contains any Python package directories.

    The way it determines this is by checking for */__init__.py* files.
    """
    for _ in glob.iglob(os.path.join(path, '*', '__init__.py*')):
        return True
    return False

def find_roots(dirpath):
    r"""Generates the root directories to add to sys.path.

    The root directories are generated bottom to top and include the
    bottom-most directory if it's not a package, then all higher directories
    that are not packages themselves, but contain packages.

    dirpath may be either the path to a file or directory.  If it is the path
    to a file, the search starts at the directory containing that file.
    """
    dirpath = os.path.normpath(os.path.abspath(dirpath))
    if not os.path.isdir(dirpath):
        dirpath = os.path.dirname(dirpath)
    lastpath = None
    while dirpath != lastpath:
        # Find first directory without an __init__.py* file.
        while dirpath != lastpath and is_package(dirpath):
            lastpath = dirpath
            dirpath = os.path.dirname(dirpath)
        # dirpath == lastpath or not is_package(dirpath)
        if dirpath == lastpath:
            break
        # not is_package(dirpath)

        # If the first directory was not a package directory, we want it
        # included!
        # Otherwise, we know that this has packages, because we just left one!
        yield dirpath

        lastpath = dirpath
        dirpath = os.path.dirname(dirpath)
        while dirpath != lastpath and not is_package(dirpath):
            if has_package(dirpath):
                # This covers the case where a directory includes packages
                # other than the sub-directory we just came from.
                yield dirpath
            lastpath = dirpath
            dirpath = os.path.dirname(dirpath)
        # dirpath == lastpath or is_package(dirpath)

def setpath(filepath, remove_cwd = True, remove_first = False, full = False):
    r"""Add the appropriate prefix of filepath to sys.path.

    This searches backwards up the list of directories in filepath looking for
    the first one that is not a Python package directory (i.e., does not
    contain an __init__.py file).  It then adds this final directory to
    Python's path (in sys.path).

    If remove_first is True, it will remove the first entry on sys.path.  This
    is used when called from a script.  When the script is run as "python
    somewhere/foobar.py", python adds the directory containing foobar.py to
    sys.path.  The remove_first option undoes this (albeit blindly -- it
    doesn't do any sanity checks first).  But note that python does _not_ add
    anything to sys.path when called with the '-m' option, for example:
    "python -m somewhere.foobar".  In this case, "somewhere" must have been on
    the python path to begin with for python to even find it in the first
    place...

    If remove_cwd is True, it will remove an '' entry (if any) from sys.path.

    If full is True, the search is done all the way to root adding all
    directories containing packages to sys.path.  (Note that if the first
    directory is not a package, it is also added to sys.path).
    """
    if remove_first:
        # kill path to this program...
        #sys.stderr.write("removing %r from sys.path\n" % sys.path[0])
        del sys.path[0]
    if remove_cwd and '' in sys.path: sys.path.remove('')
    paths = []
    for dirpath in find_roots(filepath):
        if dirpath in sys.path: sys.path.remove(dirpath)
        paths.append(dirpath)
        if not full: break
    sys.path[0:0] = paths
    return paths



doctest_tools/testdoc.py
#!/usr/bin/python

# testdoc.py [-d] [-r] file

import sys
import os, os.path
import doctest
import warnings
from doctest_tools import setpath

debug = False

warnings.simplefilter('default')

def import_module(modulepath, remove_first_path = False, full = True):
    r"""Imports the module indicated by modulepath.

    Also adds the proper containing directories to Python's sys.path.

    Returns the imported module.
    """
    pythonpath = \
      setpath.setpath(modulepath, remove_first=remove_first_path, full=full)
    if debug:
        sys.stderr.write("setpath added: %s\n" % (pythonpath,))
    modulepath = modulepath[len(pythonpath[0]) + 1:]
    if debug:
        sys.stderr.write("modulepath: %s\n" % (modulepath,))
    modulename = modulepath.replace('/', '.').replace(os.path.sep, '.')
    if debug:
        sys.stderr.write("modulename: %s\n" % (modulename,))
    module = __import__(modulename)
    for comp in modulename.split('.')[1:]:
        module = getattr(module, comp)
    return module

def test(path, remove_first_path = False, full = True):
    r"""Runs doctest on the file indicated by 'path'.

    This will run testmod if the file ends in .py, .pyc or .pyo; and testfile
    for all other files.

    When running testfile on python 2.5, it enables Python's "with" statement
    (as if the file being tested had done "from __future__ import
    with_statement").  This is done because doing the __future__ import does
    not work in files. :-(

    Also when running testfile, the current working directory is first set to
    the directory containing the file.  This is not done for python modules
    (.py, .pyc or .pyo files).

    In all cases, all non-package directories containing package directories
    (i.e., directories containing an __init__.{py,pyc,pyo} file) are added to
    sys.path.  The search is started in the directory containing the file.  If
    the bottom-most directory is not a package directory, it is added to the
    path too.
    """
    path = os.path.normpath(path)
    fullpath = os.path.abspath(path)
    if path.endswith('.py'):
        module = import_module(fullpath[:-3], remove_first_path, full)
    elif path.endswith(('.pyc', '.pyo')):
        module = import_module(fullpath[:-4], remove_first_path, full)
    else:
        new_paths = \
          setpath.setpath(fullpath, remove_first=remove_first_path, full=full)
        if debug:
            sys.stderr.write("setpath added: %s\n" % (new_paths,))
        os.chdir(os.path.dirname(fullpath))
        if sys.version_info[:2] == (2, 5):
            import __future__
            return doctest.testfile(fullpath, False,
                                    globs={
                                      'with_statement':
                                        __future__.with_statement,
                                    })
        else:
            return doctest.testfile(fullpath, False)
    module.doing_doctest = True
    return doctest.testmod(module)

def usage():
    sys.stderr.write("usage: %s [-d] [-r] file\n"
                       "       if -d is specified, debug output is turned on\n"
                       "       if -r is specified, the number of errors "
                       "and tests is printed to stderr\n"
                     % os.path.basename(sys.argv[0]))
    sys.exit(2)

def run_command(remove_first_path = False):
    r"""Process the command line args and call test().

    If the '-r' option is given, the number of errors and tests is printed to
    stdout separated by a space.

    Returns an exit status of 1 if any errors are reported.
    """
    global debug
    print_numbers = False
    command_args = sys.argv[1:]
    if not command_args:
        usage()
    if command_args[0] in ('-h', '--help'):
        usage()
    if command_args[0] == '-d':
        if len(command_args) < 2:
            usage()
        debug = True
        del command_args[0]
    if command_args[0] == '-r':
        if len(command_args) < 2:
            usage()
        print_numbers = True
        del command_args[0]
    if len(command_args) != 1:
        usage()
    filename = command_args[0]
    try:
        errors, tests = test(filename, remove_first_path)
    except IOError:
        sys.stdout.write("%s: file not found\n\n" % filename)
        usage()
    except ImportError:
        sys.stdout.write("%s: module not found\n\n" % filename)
        usage()
    if print_numbers:
        sys.stdout.write("TESTDOC RESULTS: Errors %d, Tests %d\n" %
                           (errors, tests))
    if errors: sys.exit(1)

if __name__ == "__main__":
    run_command()


doctest_tools/testall.py
#!/usr/bin/python

# testall.py [-3] [suffix...]

import os, os.path
import sys
import subprocess
import traceback
import optparse
import fnmatch
import re

def execute(args, line_filter = None):
    r"""Executes args using subprocess and prints its output.

    The output printed includes stdout and stderr.

    If line_filter is not None, it is applied against each line.  If it
    returns None, that line is printed to stdout.  If it returns anything
    else, that line is not printed and a list of the accumulated non-None
    values is returned.

    Returns a two tuple:
        - the list of non-None results from line_filter
        - the process return code.
    """
    child = subprocess.Popen(args,
                             stdin=None,
                             stdout=subprocess.PIPE,
                             stderr=subprocess.STDOUT)
    out = child.communicate()[0]
    if not isinstance(out, str): out = out.decode()  # for python3
    #sys.stdout.write("communicate got ")
    #sys.stdout.write(repr(out))
    #sys.stdout.write(" end output\n")
    lines = out.split('\n')
    if lines and not lines[-1]: del lines[-1]
    #sys.stdout.write("lines: %r\n" % lines)
    ans = []
    for line in lines:
        filtered = line_filter and line_filter(line)
        if filtered is None:
            sys.stdout.write(line + '\n')
        else:
            ans.append(filtered)
    return ans, child.returncode

testdoc_results_re = re.compile(r'TESTDOC RESULTS: Errors ([0-9]+), Tests ([0-9]+)\r?$')

def call_testdoc(path, py3kwarning = True):
    r"""Calls testdoc in a subprocess.

    The version of python used to run testdoc is the same as the version of
    python running this function.

    The stdout and stderr from testdoc are written to stdout.

    The number of errors and tests is returned as a 2-tuple.

    The py3kwarning flag will turn on the '-3' option for python if python 2.6+
    is being called.  This outputs warnings about constructs that are not
    easily translated to Python 3 with the Python's 2to3 tool.  The option is
    ignored if running Python 2.5 or before, or Python 3.
    """
    if py3kwarning and sys.version_info[0] == 2 and sys.version_info[1] >= 6:
        matches, status = execute((sys.executable, '-3',
                                     '-m', 'doctest_tools.testdoc',
                                     '-r', path),
                                  testdoc_results_re.match)
    elif sys.version_info[:2] >= (2, 5):
        matches, status = execute((sys.executable,
                                     '-m', 'doctest_tools.testdoc',
                                     '-r', path),
                                  testdoc_results_re.match)
    else:
        matches, status = execute((sys.executable,
                                     '-c',
                                     'from doctest_tools import testdoc; '
                                       'testdoc.run_command()',
                                     '-r', path),
                                  testdoc_results_re.match)

    if matches:
        try:
            return tuple(int(x) for x in matches[0].groups())
        except ValueError:
            pass
    else:
        sys.stdout.write('ERROR: testdoc failed for %r, return code %d\n' %
                           (path, status))
    return 1, 0

def filename_key(filename):
    r"""somepath.suffix => (suffix, somepath)
    
    This is only used as the 'key' function to sort.
    """
    base, ext = os.path.splitext(filename)
    return ext, base

def read_args(dirpath, dict):
    r"""Read the testall.config file in dirpath into dict.
    """
    f = open(os.path.join(dirpath, 'testall.config'))
    try:
        for line in f:
            line = line.strip()
            if not line or line[0] == '#': continue
            args = line.split()
            dict.setdefault(args[0], []).extend(args[1:])
    finally:
        f.close()

def include(option_stack, name, option_type = '', default = False):
    r"""Tests whether name is included by option_type in option_stack.

    Looks through option_stack for first occurance of a glob pattern matching
    name in 'exclude' + option_type or 'include' + option_type.  A hyphen is
    added to the front of option_type, unless it's ''.
    
    Returns:
        - False if name found in 'exclude' + option_type list
        - True if name found in 'include' + option_type list
        - otherwise default
    """
    #print "include", name, repr(option_type), default
    def name_matches(prefix):
        #print "name_matches", name, prefix + option_type, options
        for pattern in options.get(prefix + option_type, ()):
            #print "name_matches testing", name, pattern
            if fnmatch.fnmatch(name, pattern):
                #print "name_matches matched, returning True"
                return True
        #print "name_matches not found, returning False"
        return False
    for dirpath, options in option_stack:
        if name_matches('exclude'):
            #print "matches exclude, returning False"
            return False
        if name_matches('include'):
            #print "matches include, returning True"
            return True
    #print "no match, returning", default
    return default

def run(suffixes, py3kwarning = False):
    r"""Recursively look for files and run testdoc on them.

    This starts in the current working directory and recursively looks for all
    files ending in one of the 'suffixes' provided (default 'py', 'tst', and
    'txt').  It runs testdoc on each of the files found in a separate process
    (so that tests don't cross contaminate each other).

    It automatically ignores the following directories: .hg, .svn, build and
    dest.  This list is currently hard-coded...

    It also automatically ignores any file starting with 'setup' and ending
    with '.py'.  This is also currently hard-coded...

    Finally, if there is a file called 'testall.exclude' in any subdirectory,
    the file is read to get a list of names (one per line, with lines starting
    with '#' ignored).  These names can be either directory names or file
    names.  Globbing is not allowed.  All of the names listed will be ignored
    within this directory.  BUT, note that these names will _not_ be ignored
    within subdirectories!

    The order that the tests are run is:
    
      - all matching files sorted first by file extension, then by file name.
      - all subdirectories, in sorted order.

    This function returns a 4-tuple:
    
      - the number of files tested
      - the number of tests run within those files
      - the number of tests that had errors
      - a list of the paths names for the files that had errors.
    """
    #sys.stdout.write("run %r\n" % suffixes)

    errors = 0
    tests = 0
    files = 0
    error_files = []
    option_stack = []

    for dirpath, dirnames, filenames in os.walk('.'):
        if '.hg' in dirnames: dirnames.remove('.hg')
        if '.svn' in dirnames: dirnames.remove('.svn')
        if 'build' in dirnames: dirnames.remove('build')
        if 'dist' in dirnames: dirnames.remove('dist')

        #print "dirpath", dirpath
        #print "option_stack before", option_stack
        if dirpath == '.':
            # list of (dirpath, [glob_pattern])
            option_stack = [('.', {'include-suffix':
                                     suffixes or ('py', 'tst', 'txt')}
                            )]
        else:
            parent = os.path.split(dirpath)[0]
            while option_stack and option_stack[0][0] != parent:
                del option_stack[0]
            assert option_stack and option_stack[0][0] == parent, \
                   "logic error in option_stack unwinding, couldn't find %r" % \
                     parent
            option_stack.insert(0, (dirpath, {}))

        if 'testall.config' in filenames:
            read_args(dirpath, option_stack[0][1])
        #print "option_stack after", option_stack

        for dir in dirnames[:]:
            if not include(option_stack, dir, default=True):
                dirnames.remove(dir)

        dirnames.sort()

        for filename in sorted(filenames, key=filename_key):
            if filename.startswith('setup') and filename.endswith('.py'):
                continue
            suffix = os.path.splitext(filename)[1]
            if suffix: suffix = suffix[1:]  # drop the leading '.'
            if include(option_stack, filename,
                           default=include(option_stack, suffix, '-suffix')):
                path = os.path.join(dirpath, filename)[2:]
                sys.stdout.write("Testing %s\n" % path)
                files += 1
                try:
                    e, t = call_testdoc(os.path.abspath(path), py3kwarning)
                except Exception:
                    e, t = 1, 0
                    traceback.print_exc()
                if e: error_files.append(path)
            else:
                e = t = 0
            errors += e
            tests += t
    return files, tests, errors, error_files


def run_command():
    r"""Process command line args and call run().

    This also prints (to stdout) a summary of the number of files, tests and
    errors encountered, along with a list of the files that had errors.

    Returns an exit status of 1 if any errors are reported.
    """
    parser = optparse.OptionParser(
               usage="usage: %s [options] [suffix...]" %
                       os.path.basename(sys.argv[0]))
    parser.add_option('-3', action='store_true', dest='py3kwarning',
                            default=False,
                            help="warn about Python 3.x incompatibilities "
                                 "that 2to3 cannot trivially fix")
    parser.add_option('-s', '--summary', 
                            metavar="FILENAME",
                            help="file to write summary counts and list of "
                                 "error files to (stdout)")

    options, args = parser.parse_args()

    files, tests, errors, error_files = run(args, options.py3kwarning)

    if not options.summary or options.summary == 'stdout':
        f = sys.stdout
        f.write("\n")
        do_close = False
    elif options.summary == 'stderr':
        f = sys.stderr
        do_close = False
    else:
        f = open(options.summary, 'w')
        do_close = True
    f.write("Files: %d, Tests: %d, Errors: %d\n" % (files, tests, errors))
    if errors:
        f.write("********** ERRORS ************* "
                "%d files had errors:\n" % len(error_files))
        for fn in error_files:
            f.write(fn + '\n')
    if do_close: f.close()
    if errors: sys.exit(1)

if __name__ == "__main__":
    run_command()
